### ISSUE TITLE
[Enhancement]: mass converting to m4b

### ISSUE THREAD

**Issue by Blindie92** [advplyr/audiobookshelf#1267]
### Describe the feature/enhancement

i hope it is possible to make the converting tool global, instead for each book individual. I mean that i didnt need to navigate to each book and select the tool individual, instead a tool where i can multi select books i want to convert should be nice

**Comment by andrewkomarow** [comment]
Yes, whether imported or scanned. Also ability to to mass update metadata after too. Even chapters if the match is close enough.

**Comment by ReaderGuy42** [comment]
Any updates? This would be really useful.

**Comment by manderso7** [comment]
Yes, I think this would be a very cool feature. There are many apps out there that do this, none very intuitive/automated. This would solve the intuitive issue and having a multi-select/auto-convert uploads option would be ideal.

**Comment by xcy7e** [comment]
I built a python script that allows you to **start m4b converting for a single book** using selenium.
You can call the script with a `libraryItemId` (I use a cronjob that starts a bash-script that collects the next `libraryItemId` from a txt-file)

The bash-script counts the running `ffmpeg`-processes (one uses 1 cpu) and starts a new m4b-convert process through the python script, if there is enough cpu's left.
In the night, I keep a 2-hour window to allow the recent m4b-processes to finish, to then clear the metadata-cache items (using another cronjob), at the end of the 2-hour shift - to avoid my harddisk being filled up.

Hope this is useful to some of you:
https://github.com/xcy7e/Python-Scripts/blob/master/abs/abs_m4bconvert.py

**Comment by pivoivo** [comment]
Since the UI changed I asked chatgpt to help me adjust the script to work again with the new advanced tab. So far its doing well. @xcy7e maybe you could take a look and add it to you repo.

```
#
# Script to log into ABS using selenium and click the m4b convert button for a single book
#

# USAGE:
# python3 abs_m4bconvert.py {libraryItemId}
# python3 abs_m4bconvert.py "9ae4c7bb-f8d0-47dc-ae7f-6218ba479c34"

abs_host="http://YOUR_IP_OR_URL/audiobookshelf"   
abs_username="PASSWORD"
abs_password="USERNAME"
abs_convert_btn_text="M4B-Kodierung starten "   # Be aware of a trailing space in the german language!

import sys
import time

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

from selenium.webdriver.firefox.options import Options as FirefoxOptions


browser = None

try:
    opt = FirefoxOptions()
    opt.add_argument("--headless")

    browser = webdriver.Firefox(options=opt)

    bookId = sys.argv[1]
    print("Start session for book {" + bookId + "}")

    # 1. Login
    url = abs_host + "/login/?redirect=%2Faudiobook%2F" + bookId + "%2Fmanage%3Ftool%3Dm4b"

    browser.get(url)

    print("CURRENT TITLE:", browser.title)
    print("CURRENT URL:", browser.current_url)

    wait = WebDriverWait(browser, 10)

    input_element = wait.until(
        EC.presence_of_element_located((By.NAME, "username"))
    )

    usernameInput = browser.find_element(By.NAME, "username")
    passwordInput = browser.find_element(By.NAME, "password")

    usernameInput.send_keys(abs_username)
    passwordInput.send_keys(abs_password + Keys.RETURN)

    print("logging in")

    WebDriverWait(browser, 15).until(
        EC.url_changes(url)
    )

    # 2. M4B Convert page
    new_url = abs_host + "/audiobook/" + bookId + "/manage?tool=m4b"
    embed_url = abs_host + "/audiobook/" + bookId + "/manage?tool=embed"

    print(browser.current_url)

    if browser.current_url != new_url:
        if browser.current_url == embed_url:
            raise Exception("Book is already converted to m4b. Exit.")
        else:
            raise Exception(
                "Unknown redirection to [" + browser.current_url + "]. Exit."
            )

    assert "Audiobookshelf" in browser.title

    print("Convert-Page loaded")


    # 3. Open Advanced section
    advancedBtn = wait.until(
        EC.element_to_be_clickable(
            (By.XPATH, "//button[.//p[contains(text(),'Advanced')]]")
        )
    )

    advancedBtn.click()

    print("Advanced opened")



    # 4. Find and click Start M4B Encode button
    m4bBtn = wait.until(
        EC.element_to_be_clickable(
            (By.XPATH, "//button[contains(.,'Start M4B Encode')]")
        )
    )

    print("Convert-Button found")

    print("Start converting")

    m4bBtn.click()

    print("Converting started!")

    time.sleep(10)

    print("finish")

except AssertionError as ae:
    print("Assertion failed:", ae)

except Exception as error:
    print("Error:", error)

finally:
    if browser:
        browser.quit()

```

**Comment by xcy7e** [comment]
Thanks! @pivoivo

**Comment by Vito0912** [comment]
Is there a reason you do that with Selenium? Doing that with Selenium is not only extremely prune to fail, extremely ineffecient, but it will also fail in a few months, because ABS completely reworks their UI in the upcoming month(s).

You also have your password somewhere in plain text.

You can simply make a POST request (below just an illustrative example):
```
            await this.client.post(
                `/api/tools/item/${itemId}/encode-m4b?codec=${codec}&bitrate=${bitrate}&channels=${channels}`,
                {},
            );
```

**Comment by pivoivo** [comment]
> Is there a reason you do that with Selenium? Doing that with Selenium is not only extremely prune to fail, extremely ineffecient, but it will also fail in a few months, because ABS completely reworks their UI in the upcoming month(s).
> 
> You also have your password somewhere in plain text.
> 
> You can simply make a POST request (below just an illustrative example):
> 
> ```
>             await this.client.post(
>                 `/api/tools/item/${itemId}/encode-m4b?codec=${codec}&bitrate=${bitrate}&channels=${channels}`,
>                 {},
>             );
> ```

I am just not able to do this. For now the solutions work. But if you want to make something which is more reliable with further updates I would be happy.

**Comment by xcy7e** [comment]
@Vito0912 Yes, you’re right. This was a one-time use case for me. I’ve since caught up with converting my library from MP3 to M4B, and now I manually convert each book as soon as I add it.

I shared the script because the OP asked for a feature I had just implemented. Since there seems to be interest from more people, it would definitely make sense to turn it into a more robust and maintainable solution.

My own use case has been solved for quite some time, so I don’t currently plan to spend more time on it myself, though maybe I'll have some free time in the future, to work in this. That said, I’d be happy if someone wanted to build on the idea or take it further.

**Comment by Vito0912** [comment]
> I am just not able to do this. For now the solutions work. But if you want to make something which is more reliable with further updates I would be happy.

I already have an [abs-autoconverter](https://github.com/Vito0912/abs-autoconverter), which watches and converts for changes.

There is also this [auto converter](https://github.com/cutzenfriend/docker-absautoconverter). For a long time it was unmaintained and lacked a lot of features, but got some new features lately

(It might not be what you need exactly though)

> I shared the script because the OP asked for a feature I had just implemented. Since there seems to be interest from more people, it would definitely make sense to turn it into a more robust and maintainable solution.

All good. I do not say how people should do things. It was just to make aware that there is an API. Because the client also just calls this exact API if the button gets triggered

**Comment by gavinmcfall** [comment]
For anyone wanting batch conversion today without waiting on a UI feature, here's the approach I just used to convert ~60 books (mostly 1GB+ audiobooks) by driving the same per-item API from a small script. The API call itself is the easy part — the lessons below are what actually matter at batch scale:

**The loop:**
1. Enumerate library folders that contain mp3/m4a and no m4b (that check is what makes the script resumable — kill it any time and relaunch, finished books are skipped). Sort smallest-first so you bank quick wins early.
2. `POST /api/tools/item/{libraryItemId}/encode-m4b`, then poll the tasks API until the encode finishes.
3. Verify, purge the cache for that item, move to the next book. Strictly one at a time.

**Hard-won details:**

- **Run sequentially, never in parallel.** Parallel encodes/scans OOMKilled my ABS container (1Gi limit at the time). One encode at a time was completely stable.
- **Verify success on disk, not from the item record.** Right after an encode the item record can lag behind reality — I got a false "failure" that way. The reliable check is the folder itself: exactly 1 `.m4b` present, 0 leftover mp3/m4a, sane file size.
- **Watch `/config/metadata/cache/items/{id}`.** ABS moves the original mp3s there rather than deleting them (good safety net!), but at batch scale that's a silent disk eater — 60 books × ~1GB would have blown through my config volume. I purged each book's cache dir only after the disk verification passed.
- **Dual-format folders:** the encoder writes the m4b to the title-folder root. If you keep ebook+audio together with audio in a subfolder, you'll want a post-pass step to file the m4bs where you want them, then rescan.
- **Stagger your rescans.** Same OOM reason — I rescanned affected libraries one at a time at the end of the pass rather than after every book.

Throughput for me was roughly 35–55 MB of source audio per minute per book on modest hardware, fully unattended.

One failure mode I haven't cracked: a single 101-track book returns from the encode task with no m4b produced, three attempts in a row, no error surfaced — everything else (up to 116 tracks) converted fine. I'll convert that one outside ABS.

+1 to the feature itself though — a multi-select in the UI doing exactly this loop (sequential queue + cache cleanup) would remove the need for scripts entirely.
